### ISSUE TITLE
Deprecate NeDB in favour of AceBase

### ISSUE THREAD

**Issue by routerino** [Ylianst/MeshCentral#4398]
[NeDB](https://github.com/louischatriot/nedb) is no longer supported and hasn't been supported for some time, but it is the default embedded database (I'm assuming due to its compatibility with mongodb).

[Acebase](https://github.com/appy-one/acebase) is a current mongodb compatible embedded database that uses (among other options) sqlite as the backend storage, meaning that (theoretically) the proven performance and reliability of sqlite can be leveraged with mongodb syntax queries.

Of course NeDB can be considered "good enough" and serious use should switch to mongo anyway, but it would be nice not to rely on an unmaintained package for the core database.

**Comment by Ylianst** [comment]
I agree with this. The benefit of NeDB is that it was "Pure NodeJS" and so, there was no risk for failing to install on any platform. I was also sort-of compatible with the MongoDB API. This said, a lot of people use MeshCentral with the embedded database and so, getting something more robust would be welcome.

**Comment by Ylianst** [comment]
Well, that took my entire Saturday, but I added basic support for AceBase. As expected, it's a local database with no dependencies. There is plenty more work to do to test it and to improve performance, but the basics seem to work. I am not going to make it the default database for a while, but we can start testing it.

![MC2-AceBaseSupport](https://user-images.githubusercontent.com/1319013/184519108-aa5a8e26-219b-4461-9a5c-96b8d09a86ab.png)

This will be in MeshCentral 1.0.66 when it next comes out.

**Comment by Ylianst** [comment]
So, an update on this topic. I spent the weekend adding AceBase and at first it looked good, it's a pure NodeJS solution and it's supported. However, after running it for a while on my development server I started noticing issues. It does not handle corrupt data well and removing records from the database it terribly slow. Removing 1000 old records in the events table can lock up the server for over a minute... like REALLY slow. It looks like each time I remove a record, it rebuilds the database indexes. The indexes are also huge. My tiny database with 2000 records (mostly events) has an index file of over 100 megabytes.

So, I am not sure I want to trust people's data to AceBase, even for small servers. I am going to keep the code there in case someone wants to use it, but I am not going to be using AceBase as the next NeDB.

I started working on SQLite support and that seems to be going well so far. It's not pure NodeJS and has a lot of dependencies, but it seems to be fast and reliable. I will likely complete support for it today and see how well that works.

**Comment by si458** [comment]
You could look into pouchdb https://pouchdb.com/
It supports local dbs, external couchdb servers and even adapters for other engines
https://pouchdb.com/adapters.html

**Comment by Ylianst** [comment]
One more to add :) I am almost done with SQLite. I will get that done first.

**Comment by Ylianst** [comment]
Just published MeshCentral v1.0.66 with AceBase support, but I changed the graphic to "trial" as I am not super happy with it. I did get SQLite working and that works a lot better.

![MC2-AceBaseSupport](https://user-images.githubusercontent.com/1319013/184748891-d160ea7b-8981-48dd-a818-a4fb23948af3.png)

**Comment by routerino** [comment]
> Just published MeshCentral v1.0.66 with AceBase support, but I changed the graphic to "trial" as I am not super happy with it. I did get SQLite working and that works a lot better.
> 
> ![MC2-AceBaseSupport](https://user-images.githubusercontent.com/1319013/184748891-d160ea7b-8981-48dd-a818-a4fb23948af3.png)

Honestly I suggested acebase because it's new and worked "well enough" for when I tinkered with it, but sqlite would be a much more battle hardened choice anyway. SQLite is the "de facto" embedded database, it's not going out of support any time soon.

**Comment by cgb** [comment]
Hey @Ylianst great to see SQLite support being offered...!  I experimented with moving to Postgresql but had to fall back to Mongodb due to plugin Scripttask relying on it (or NeDB). 

I understand that plugins are not officially supported, but it would be great if we could work out between @ryanblenis, Meshcentral developers and community how plugins can reliably interact with the database without needing db-specific handling (as is the case currently in Scriptask in db.js).

**Comment by Ylianst** [comment]
I really just have to bite the bullet and merge Scriptask in MeshCentral. It's used by many and I have to make it something that is officially supported.

**Comment by Ylianst** [comment]
Wow, I love how the first thing on [Hacker news](https://news.ycombinator.com/) right now is: [SQLite is not a toy database](https://antonz.org/sqlite-is-not-a-toy-database/).

**Comment by cgb** [comment]
Yeah SQLite has been getting a lot of airtime on HN lately, specifically Fly.io's blog items.

https://fly.io/blog/all-in-on-sqlite-litestream/
https://fly.io/blog/sqlite-internals-btree/
https://fly.io/blog/sqlite-internals-rollback-journal/

And some projects working on replicating SQLite benath the DB:

https://litestream.io/
https://github.com/superfly/litefs

Exciting times to begin using SQLite :)

**Comment by Ylianst** [comment]
In having enabled AceBase and SQLite within the same week, I can tell just how much better and more reliable SQLite is. The airtime is certainly justified.

**Comment by appy-one** [comment]
@Ylianst can you submit the issues you ran into with AceBase so I can take a look at them? Huge indexes should only be there if the amount of data indexed is huge, there is not much overhead in the index files. Also, deletion of 1000s of records should be fairly quick, so I'm curious how you are using the db. Kindly note that every database engine has the potential to be quick and slow at the same time, based upon how it's used and the optimisations in place. Would really love to help iron out any issues you ran into!

**Comment by si458** [comment]
@appy-one @Ylianst is currently away on vacation, https://github.com/Ylianst/MeshCentral/discussions/4460
he will reply as and when he can 👍

**Comment by marclaporte** [comment]
So.... can we close this as done?

And issues/concerns about the SQLite implementation can be managed via other issues or discussions?
https://github.com/Ylianst/MeshCentral/discussions/5532

**Comment by PTR-inc** [comment]
Just did a quick look into the acebase functionality in meshcentral, but at the moment it does not seem to be working at all.
There is a missing escapefieldname for the mesh type in the get/storeevents function and some functionparameters seem to have changed, or made more strict? f.e. https://github.com/Ylianst/MeshCentral/blob/e58d659fa94a74e6bbbef5198f85ec10e8a3993e/db.js#L1935
the .filter throws an error for the 'in' filter expecting an array and getting a string.
Did some quickfixes, but still not working 100%

My opinion in this case: go for sqlite as the default local filedatabase. There is only need for one local db I would say. From a maintenance viewpoint.

**Comment by PTR-inc** [comment]
@si458 In light of the recent nedb/sqlite discussion #6503 #5532  and this one, is it an idea to add sqlite config options to be able to adjust journalling and performance? I could make a pr for it today, it's mostly adding pragma's at startup.

**Comment by si458** [comment]
@PTR-inc thats perfectly fine, if you think you can add extra options which users can toggle to do things to help improve sqlite then yes have a try at it and see how you get on!

**Comment by NealVanMeerbeekPXL** [comment]
Just to add my 2 cents: this fork of NeDB is well-maintained and used as a replacement by e.g. Insomnia.REST: https://github.com/seald/nedb

**Comment by si458** [comment]
@NealVanMeerbeekPXL we already moved to that fork of NeDB a few releases ago 👍 1.1.34
https://github.com/Ylianst/MeshCentral/commit/975e49a1905a9c9559ea09bbd7907867f457144a

**Comment by marclaporte** [comment]
> we already moved to that fork of NeDB

So this is done? (vs not planned)

**Comment by si458** [comment]
@Ylianst agrees we should move from NeDB to SQLite, 
but we are waiting for SQLite to be fully integrated into NodeJS first,
22+ has a flag u can enable to support SQLite but its still in Active development
https://nodejs.org/docs/latest-v22.x/api/sqlite.html#sqlite

where as at the moment for sqlite and other database engines we have to install npm modules to support them
and with sqlite it has to BUILD nan packages/binarys which are based on the OS its being installed on so its a nightmare!

happy for you to reopen it but it will just take time to move from NeDB maybe next year i think 👍
